File: app/src/main/java/net/squanchy/support/graphics/CircularRevealDrawable.java
package net.squanchy.support.graphics;

import android.animation.Animator;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.ColorDrawable;
import android.support.annotation.ColorInt;
import android.support.annotation.IntRange;
import android.support.v4.view.animation.FastOutLinearInInterpolator;
import android.view.animation.Interpolator;

import me.eugeniomarletti.renderthread.CanvasProperty;
import me.eugeniomarletti.renderthread.RenderThread;

public class CircularRevealDrawable extends ColorDrawable {

    private final Paint animationPaint = new Paint(Paint.ANTI_ALIAS_FLAG);

    private final Interpolator interpolator = new FastOutLinearInInterpolator();

    private Animator radiusAnimator;

    private CanvasProperty<Float> centerXProperty;
    private CanvasProperty<Float> centerYProperty;
    private CanvasProperty<Float> radiusProperty;
    private CanvasProperty<Paint> paintProperty;

    private boolean startAnimationOnNextDraw;

    private int revealDuration;

    private float hotspotX;
    private float hotspotY;
    private int width;
    private int height;

    @ColorInt
    private int pendingTargetColor;

    @ColorInt
    private int targetColor;

    public static CircularRevealDrawable from(int initialColor) {
        CircularRevealDrawable revealDrawable = new CircularRevealDrawable();
        revealDrawable.setColor(initialColor);
        revealDrawable.targetColor = initialColor;
        return revealDrawable;
    }

    public void animateToColor(@ColorInt int newColor, @IntRange(from = 0) int durationMillis) {
        revealDuration = durationMillis;
        pendingTargetColor = newColor;
        startAnimationOnNextDraw = true;
        setColor(pendingTargetColor);
    }

    @Override
    public void draw(Canvas canvas) {
        // 1. Draw the "current" color
        canvas.drawColor(getColor());

        // 2. If we just got a reveal start request, go with it
        if (startAnimationOnNextDraw) {
            initializeAnimation(canvas, pendingTargetColor);
            radiusAnimator.start();
            startAnimationOnNextDraw = false;
        }

        // 3. This draws the reveal, if one is needed
        if (radiusAnimator != null && radiusAnimator.isRunning()) {
            RenderThread.drawCircle(canvas, centerXProperty, centerYProperty, radiusProperty, paintProperty);
        }
    }

    private void initializeAnimation(Canvas canvas, int pendingTargetColor) {
        float initialRadius = 0f;
        float targetRadius = calculateTargetRadius();
        animationPaint.setColor(pendingTargetColor);

        centerXProperty = RenderThread.createCanvasProperty(canvas, hotspotX);
        centerYProperty = RenderThread.createCanvasProperty(canvas, hotspotY);
        radiusProperty = RenderThread.createCanvasProperty(canvas, initialRadius);
        paintProperty = RenderThread.createCanvasProperty(canvas, animationPaint);

        cancelTransitions();

        radiusAnimator = RenderThread.createFloatAnimator(this, canvas, radiusProperty, targetRadius);
        radiusAnimator.setInterpolator(interpolator);
        radiusAnimator.setDuration(revealDuration);
        targetColor = pendingTargetColor;
    }

    private float calculateTargetRadius() {
        float maxXDistance = Math.max(hotspotX, width - hotspotX);
        float maxYDistance = Math.max(hotspotY, height - hotspotY);
        return (float) Math.sqrt(maxXDistance * maxXDistance + maxYDistance * maxYDistance);
    }

    @Override
    public void setHotspot(float x, float y) {
        super.setHotspot(x, y);
        hotspotX = x;
        hotspotY = y;
    }

    @Override
    protected void onBoundsChange(Rect bounds) {
        super.onBoundsChange(bounds);
        width = bounds.width();
        height = bounds.height();
    }

    @Override
    public void setColor(int color) {
        targetColor = color;
        super.setColor(color);
    }

    public void cancelTransitions() {
        if (radiusAnimator == null || !radiusAnimator.isRunning()) {
            return;
        }

        radiusAnimator.cancel();
        setColor(targetColor);
    }
}

File: app/src/main/java/net/squanchy/support/graphics/CircularRevealDrawable.kt
package net.squanchy.support.graphics

import android.animation.Animator
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.drawable.ColorDrawable
import androidx.annotation.ColorInt
import androidx.annotation.IntRange
import androidx.core.animation.doOnEnd
import androidx.interpolator.view.animation.FastOutLinearInInterpolator
import me.eugeniomarletti.renderthread.CanvasProperty
import me.eugeniomarletti.renderthread.RenderThread
import kotlin.math.max
import kotlin.math.sqrt

class CircularRevealDrawable : ColorDrawable() {

    private val animationPaint = Paint(Paint.ANTI_ALIAS_FLAG)

    private val interpolator = FastOutLinearInInterpolator()

    private lateinit var radiusAnimator: Animator

    private lateinit var centerXProperty: CanvasProperty<Float>
    private lateinit var centerYProperty: CanvasProperty<Float>
    private lateinit var radiusProperty: CanvasProperty<Float>
    private lateinit var paintProperty: CanvasProperty<Paint>

    private var startAnimationOnNextDraw: Boolean = false

    private var revealDuration: Int = 0

    private var hotspotX: Float = 0f
    private var hotspotY: Float = 0f
    private var width: Int = 0
    private var height: Int = 0

    @ColorInt
    private var pendingTargetColor: Int = 0

    @ColorInt
    private var targetColor: Int = 0

    fun animateToColor(@ColorInt newColor: Int, @IntRange(from = 0) durationMillis: Int) {
        revealDuration = durationMillis
        pendingTargetColor = newColor
        startAnimationOnNextDraw = true
        color = targetColor
    }

    override fun draw(canvas: Canvas) {
        // 1. Draw the "current" color
        canvas.drawColor(color)

        // 2. If we just got a reveal start request, go with it
        if (startAnimationOnNextDraw) {
            initializeAnimation(canvas, pendingTargetColor)
            radiusAnimator.start()
            startAnimationOnNextDraw = false
        }

        // 3. This draws the reveal, if one is needed
        if (::radiusAnimator.isInitialized && radiusAnimator.isRunning) {
            RenderThread.drawCircle(canvas, centerXProperty, centerYProperty, radiusProperty, paintProperty)
        }
    }

    private fun initializeAnimation(canvas: Canvas, pendingTargetColor: Int) {
        val initialRadius = 0f
        val targetRadius = calculateTargetRadius()
        animationPaint.color = pendingTargetColor

        centerXProperty = RenderThread.createCanvasProperty(canvas, hotspotX)
        centerYProperty = RenderThread.createCanvasProperty(canvas, hotspotY)
        radiusProperty = RenderThread.createCanvasProperty(canvas, initialRadius)
        paintProperty = RenderThread.createCanvasProperty(canvas, animationPaint)

        cancelTransitions()

        radiusAnimator = RenderThread.createFloatAnimator(this, canvas, radiusProperty, targetRadius)
        radiusAnimator.interpolator = interpolator
        radiusAnimator.duration = revealDuration.toLong()
        radiusAnimator.doOnEnd {
            color = targetColor
        }

        targetColor = pendingTargetColor
    }

    private fun calculateTargetRadius(): Float {
        val maxXDistance = max(hotspotX, width - hotspotX)
        val maxYDistance = max(hotspotY, height - hotspotY)
        return sqrt(maxXDistance * maxXDistance + maxYDistance * maxYDistance)
    }

    override fun setHotspot(x: Float, y: Float) {
        super.setHotspot(x, y)
        hotspotX = x
        hotspotY = y
    }

    override fun onBoundsChange(bounds: Rect) {
        super.onBoundsChange(bounds)
        width = bounds.width()
        height = bounds.height()
    }

    override fun setColor(color: Int) {
        targetColor = color
        super.setColor(color)
    }

    fun cancelTransitions() {
        if (!::radiusAnimator.isInitialized || !radiusAnimator.isRunning) {
            return
        }

        radiusAnimator.cancel()
        color = targetColor
    }

    companion object {

        fun from(initialColor: Int) = CircularRevealDrawable().apply {
            color = initialColor
            targetColor = initialColor
        }
    }
}
